osh/string_ops.py

```diff
@@ -53,7 +55,95 @@ def _Utf8CharLen(starting_byte):
         e_strict(INVALID_START, loc.Missing)
 
 
-def _NextUtf8Char(s, i):
+def _ReadOneUnit(s, cursor):
+    # type: (str, int) -> Tuple[int, int]
+    """Helper for DecodeUtf8Char"""
+    if cursor >= len(s):
+        raise error.Expr(INCOMPLETE_CHAR, loc.Missing)
+
+    b = ord(s[cursor])
+    cursor += 1
+
+    if b & 0b11000000 != 0b10000000:
+        raise error.Expr(INVALID_CONT, loc.Missing)
+
+    return b & 0b00111111, cursor
+
+
+def DecodeUtf8Char(s, start):
+    # type: (str, int) -> int
+    """Given a string and exclusive bounds [start, end), decode the Unicode
+    char within those bounds. If the bounds must not extend beyond the size of
+    the UTF-8 codepoint, otherwise we consider the codepoint to be invalid (and
+    return -1).
+
+    The start/end bounds are in bytes and should be found using a function like
+    NextUtf8Char or PreviousUtf8Char.
+
+    If the codepoint in invalid, we raise an `error.Expr`.
+
+    Known Issues:
+    - Doesn't raise issue on surrogate pairs
+    - Doesn't raise issue on non-shortest form encodings
+    """
+    # We use table 3.6 (reproduced below) from [0]. Note that table 3.6 is not
+    # sufficient for validating UTF-8 as it allows surrogate pairs and
+    # non-shortest form encodings. A correct decoder should follow the
+    # encodings in table 3.7 from [0].
+    #
+    # | Scalar Value               | 1st Byte | 2nd Byte | 3rd Byte | 4th Byte |
+    # +----------------------------+----------+----------+----------+----------+
+    # | 00000000 0xxxxxxx          | 0xxxxxxx |          |          |          |
+    # | 00000yyy yyxxxxxx          | 110yyyyy | 10xxxxxx |          |          |
+    # | zzzzyyyy yyxxxxxx          | 1110zzzz | 10yyyyyy | 10xxxxxx |          |
+    # | 000uuuuu zzzzyyyy yyxxxxxx | 11110uuu | 10uuzzzz | 10yyyyyy | 10xxxxxx |
+    #
+    # [0] https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf
```

Comment:
Hm I just scrolled through this doc, I had never seen it

That's a doozy

Someone should write a much shorter description :)   I believe I have "all" of the UTF-8 encoding/decodnig  in my head now (not all of Unicode, but I believe the only other thing Oils needs is uppercase/lowercase)


The point of the skeleton in `data_lang/utf8_test.cc`  was to test that claim :)  By exhaustive testing

I still want to do that, although there is a lot of other stuff to do, and this PR looks like it will get us very close to where we want to be

I'm a little worried about the speed, because we're creating 1 byte strings, but correctness is more important than speed.  Decoding only the beginning and end definitely helps

Reply:
> I'm a little worried about the speed, because we're creating 1 byte strings, but correctness is more important than speed. Decoding only the beginning and end definitely helps

Yikes. I just took a look at the generated cpp, and yes, we do this. I did a simple grep for `ord(some_string->at(some_index))` and it appears we do this elsewhere too:

```
$ grep -E 'ord\(.*->at' _gen/bin/oils_for_unix.mycpp.cc
                      d = len(s) >= 2 ? ord(s->at(1)) : 0;
        delim_byte = ord(arg->d->at(0));
          tok = location::LeftTokenForWord(trail->words->at(0));
          tok = location::LeftTokenForWord(trail->words->at(-1));
    return ord(this->src_line->content->at(pos));
      start_num = ord(part->start->at(0));
      end_num = ord(part->end->at(0));
    left_tok = location::LeftTokenForWord(words->at(i));
    right_tok = location::RightTokenForWord(words->at(-1));
    left_tok = location::LeftTokenForCompoundWord(words->at(0));
  b = ord(s->at(cursor));
  b = ord(s->at(start));
  byte_as_int = ord(s->at(i));
    byte_as_int = ord(s->at(i));
      return Alloc<CharCode>(ord(tok->tval->at(0)), false, tok);
                      cval = num::ToBig(ord(consts::LookupCharC(tok_str->at(1))));
```

I guess this would be fixed by small strings? It might also be worth optimizing this into something like `int BigStr::ord_at`, which would avoid the allocation entirely.

In this case, this should be fine for now? The plan is to replace all of the python utf-8 decoding with something like the Bjorn DFA.

Reply:
Yes this is OK for now, we have that issue in other places too

Small string optimization can fix that ... I think it's feasible but not trivial, I did a little experiment in the code awhile ago

https://github.com/oilshell/oil/blob/master/mycpp/small_str_test.cc